cmd/juju/application/refresher/refresher.go

```diff
@@ -116,15 +120,34 @@ func (d *factory) maybeReadLocal(charmAdder store.CharmAdder, charmRepo CharmRep
 func (d *factory) maybeCharmStore(authorizer store.MacaroonGetter, charmAdder store.CharmAdder, charmResolver CharmResolver) func(RefresherConfig) (Refresher, error) {
 	return func(cfg RefresherConfig) (Refresher, error) {
 		return &charmStoreRefresher{
-			authorizer:     authorizer,
-			charmAdder:     charmAdder,
-			charmResolver:  charmResolver,
-			charmURL:       cfg.CharmURL,
-			charmRef:       cfg.CharmRef,
-			channel:        cfg.Channel,
-			deployedSeries: cfg.DeployedSeries,
-			force:          cfg.Force,
-			forceSeries:    cfg.ForceSeries,
+			baseRefresher: baseRefresher{
+				charmAdder:     charmAdder,
+				charmResolver:  charmResolver,
+				charmURL:       cfg.CharmURL,
+				charmRef:       cfg.CharmRef,
+				channel:        cfg.Channel,
+				deployedSeries: cfg.DeployedSeries,
+				force:          cfg.Force,
+				forceSeries:    cfg.ForceSeries,
+			},
+			authorizer: authorizer,
+		}, nil
+	}
+}
+
+func (d *factory) maybeCharmHub(charmAdder store.CharmAdder, charmResolver CharmResolver) func(RefresherConfig) (Refresher, error) {
+	return func(cfg RefresherConfig) (Refresher, error) {
+		return &charmHubRefresher{
+			baseRefresher: baseRefresher{
+				charmAdder:     charmAdder,
+				charmResolver:  charmResolver,
+				charmURL:       cfg.CharmURL,
+				charmRef:       cfg.CharmRef,
+				channel:        cfg.Channel,
```

Comment:
Can we pass an origin here instead?  To ensure the correct ID is run.  Or do you intend to change that in the follow ups?

Reply:
Follow up